p2p/src/sync/mod.rs

```diff
@@ -89,445 +97,485 @@ where
+            log::debug!("Ignoring headers request because the node is in initial block download");
+            return Ok(());
+        }
+
+        let headers = self.chainstate_handle.call(|c| c.get_headers(locator)).await??;
+        debug_assert!(headers.len() <= self.p2p_config.msg_header_count_limit.clone().into());
+        self.messaging_handle.send_message(
+            peer,
+            SyncMessage::HeaderListResponse(HeaderListResponse::new(headers)),
+        )?;
+
+        Ok(())
     }
 
-    /// Process block request
-    pub async fn process_block_request(
+    /// Processes the blocks request.
+    async fn handle_block_request(
         &mut self,
-        peer_id: T::PeerId,
-        headers: Vec<Id<Block>>,
-    ) -> crate::Result<()> {
-        ensure!(
-            headers.len() == 1,
-            P2pError::ProtocolError(ProtocolError::InvalidMessage),
-        );
-
-        // TODO: handle processing requests for multiple blocks
-        let block_id =
-            *headers.get(0).ok_or(P2pError::ProtocolError(ProtocolError::InvalidMessage))?;
-        ensure!(
-            self.peers.contains_key(&peer_id),
-            P2pError::PeerError(PeerError::PeerDoesntExist),
-        );
-
-        let block_result =
-            self.chainstate_handle.call(move |this| this.get_block(block_id)).await?;
-
-        match block_result {
-            Ok(Some(block)) => self.send_block_response(peer_id, vec![block]),
-            Ok(None) => {
-                // TODO: check if remote has already asked for these headers?
-                Err(P2pError::ProtocolError(ProtocolError::InvalidMessage))
-            }
-            Err(err) => Err(P2pError::ChainstateError(err)),
+        peer: T::PeerId,
+        mut block_ids: Vec<Id<Block>>,
+    ) -> Result<()> {
+        log::debug!("Blocks request from peer {peer}");
+
+        let peer_state = self
+            .peers
+            .get_mut(&peer)
+            .ok_or(P2pError::PeerError(PeerError::PeerDoesntExist))?;
+
+        if self.chainstate_handle.call(|c| c.is_initial_block_download()).await?? {
+            log::debug!("Ignoring blocks request because the node is in initial block download");
+            return Ok(());
+        }
+
+        let requested_blocks_limit = self.p2p_config.max_request_blocks_count.clone().into();
+        if block_ids.len() > requested_blocks_limit {
+            return Err(P2pError::ProtocolError(
+                ProtocolError::BlocksRequestLimitExceeded(block_ids.len(), requested_blocks_limit),
+            ));
         }
+        log::trace!("Requested block ids: {block_ids:#?}");
+
+        // Check that all blocks are known.
+        for id in block_ids.clone() {
+            self.chainstate_handle.call(move |c| c.get_block_index(&id)).await??.ok_or(
+                P2pError::ProtocolError(ProtocolError::UnknownBlockRequested),
+            )?;
+        }
+
+        block_ids.truncate(requested_blocks_limit - peer_state.num_blocks_to_send);
```

Comment:
Instead of checking then separately subtracting again, creating an unnecessary invariant, you can use match or if-else with an expression to extract this value or return an error.